lib/Sema/TypeCheckAttr.cpp

```diff
@@ -3466,9 +3497,88 @@ void AttributeChecker::checkOriginalDefinedInAttrs(Decl *D,
   }
 }
 
-void AttributeChecker::checkBackDeployAttrs(Decl *D,
-    ArrayRef<BackDeployAttr *> Attrs) {
-  // FIXME(backDeploy): Diagnose incompatible uses of `@_backDeploy
+void AttributeChecker::checkBackDeployAttrs(ArrayRef<BackDeployAttr *> Attrs) {
+  if (Attrs.empty())
+    return;
+
+  // Diagnose conflicting attributes. @_alwaysEmitIntoClient, @inlinable, and
+  // @_transparent all conflict with back deployment because they each cause the
+  // body of a function to be copied into the client under certain conditions
+  // and would defeat the goal of back deployment, which is to always use the
+  // ABI version of the declaration when it is available.
+  if (auto *AEICA = D->getAttrs().getAttribute<AlwaysEmitIntoClientAttr>()) {
+    diagnoseAndRemoveAttr(AEICA, diag::attr_incompatible_with_back_deploy,
+                          AEICA, D->getDescriptiveKind());
+  }
+
+  if (auto *IA = D->getAttrs().getAttribute<InlinableAttr>()) {
+    diagnoseAndRemoveAttr(IA, diag::attr_incompatible_with_back_deploy, IA,
+                          D->getDescriptiveKind());
+  }
+
+  if (auto *TA = D->getAttrs().getAttribute<TransparentAttr>()) {
+    diagnoseAndRemoveAttr(TA, diag::attr_incompatible_with_back_deploy, TA,
+                          D->getDescriptiveKind());
+  }
+
+  // @objc conflicts with back deployment since it implies dynamic dispatch.
+  if (auto *OA = D->getAttrs().getAttribute<ObjCAttr>()) {
+    diagnose(OA->getLocation(), diag::attr_incompatible_with_back_deploy, OA,
+             D->getDescriptiveKind());
+  }
+
+  // Only functions, methods, computed properties, and subscripts are
+  // back-deployable, so D should be ValueDecl.
+  auto *VD = cast<ValueDecl>(D);
+  std::map<PlatformKind, SourceLoc> seenPlatforms;
+
+  for (auto *Attr : Attrs) {
+    // Back deployment only makes sense for public declarations.
+    if (diagnoseAndRemoveAttrIfDeclIsNonPublic(Attr, /*isError=*/true))
+      continue;
+
+    // Back deployment isn't compatible with dynamic dispatch.
+    if (VD->isPotentiallyOverridable() && !VD->isFinal()) {
```

Comment:
Sorry about missing this from the previous review. Doesn't `isPotentiallyOverridable` imply non-final?

Reply:
Not quite. Here's the documentation for `isPotentiallyOverridable()`:
```
  /// Returns true if the nature of this declaration allows overrides.
  /// Note that this does not consider whether it is final or whether
  /// the class it's on is final.
```
I found in testing that I needed to check both conditions. First, we need to determine whether or not the decl is a decl where asking "is this final?" makes sense, and that is answered by `isPotentiallyOverridable()`. Then we can check whether it is in fact `final` before diagnosing.

Reply:
Thanks for the explanation! I wonder whether it makes sense to integrate `isFinal` check into `isPotentiallyOverridable()` because it's certain that final decls isn't potentially overridable.

Reply:
I think the existing semantics of `isPotentiallyOverridable()` are needed by existing call sites.

Reply:
Sure, what I'm trying to say is very likely integrating `isFinal` into `isPotentiallyOverridable()` won't change the behavior of the existing call sites, and it's a sensible thing to do for the API per se.

Even if some tests failed because of this change, it would be interesting to learn what are these unexpected situations.

Reply:
I looked at the call sites and I see a couple where making that change doesn't look safe. If I understand correctly, you wondered about this because the name `isPotentiallyOverridable()` could be interpreted to behave the way you're thinking and I agree it's a reasonable interpretation. However, I'm not sure we should change its behavior to match the name when it is documented clearly to have the behavior it does. I'd rather change the name to better match the behavior, though I have thought about this and I haven't come up with a better name.

Reply:
👍I think you are right that my confusion is mostly because the name `isPotentiallyOverridable()` doesn't indicate the purpose of this check. We don't have to update this API in this specific PR. We could later experiment with a separate one.